keps/sig-auth/1688-dynamic-authentication-config/README.md

```diff
@@ -0,0 +1,620 @@
+be used.
+
+[#71162]: https://github.com/kubernetes/kubernetes/issues/71162
+[#166]: https://github.com/aws/containers-roadmap/issues/166
+[kube-oidc-proxy]: https://aws.amazon.com/blogs/opensource/consistent-oidc-authentication-across-multiple-eks-clusters-using-kube-oidc-proxy
+[teleport]: https://aws.amazon.com/blogs/opensource/authenticating-eks-github-credentials-teleport
+[dex]: https://github.com/dexidp/dex
+
+### Goals
+
+<!--
+List the specific goals of the KEP.  What is it trying to achieve?  How will we
+know that this has succeeded?
+-->
+
+1. Create a new Kubernetes REST API that allows configuration of authentication
+  - x509
+  - OIDC
+  - Token Webhook
+2. Changes made via the REST API should be active in the order of minutes without
+requiring a restart of the API server
+
+### Non-Goals
+
+<!--
+What is out of scope for this KEP?  Listing non-goals helps to focus discussion
+and make progress.
+-->
+
+1. Changing the internal behavior of the authentication modes (ex: using
+different fields in x509 certificates to build the `user.Info` object)
+2. Restricting the REST API to the exact structure of the CLI flags
+
+## Proposal
+
+<!--
+This is where we get down to the specifics of what the proposal actually is.
+This should have enough detail that reviewers can understand exactly what
+you're proposing, but should not include things like API designs or
+implementation.  The "Design Details" section below is for the real
+nitty-gritty.
+-->
+
+This change aims to add a new API Kubernetes REST API called
+`AuthenticationConfig`.  It is similar to the `ValidatingWebhookConfiguration`
+API in that it will be watched by the API server and used to construct a struct.
+In this case, this struct with be an `authenticator.Request`, i.e. a dynamic
+authenticator.  This authenticator will be unioned with the other authenticators.
+
+### User Stories (optional)
+
+<!--
+Detail the things that people will be able to do if this KEP is implemented.
+Include as much detail as possible so that people can understand the "how" of
+the system.  The goal here is to make this feel real for users without getting
+bogged down.
+-->
+
+#### Story 1
+
+A user wishes to use Dex on an existing cluster.  They follow Dex's instructions
+to deploy the IDP onto the cluster.  They then create a matching
+`AuthenticationConfig` object with the `spec.type` set to `oidc` along with other
+relevant options.  This allows the API server to validate ID tokens issued by Dex.
+The user configures the `kubectl` OIDC integration and uses a Dex provided
+identity to authenticate to the cluster.
+
+### Notes/Constraints/Caveats (optional)
+
+<!--
+What are the caveats to the proposal?
+What are some important details that didn't come across above.
+Go in to as much detail as necessary here.
+This might be a good place to talk about core concepts and how they relate.
+-->
+
+TBD.
+
+### Risks and Mitigations
```

Comment:
This new API is obviously admin-level. Are there any gaps in RBAC to managing this API? Ex: "I only want to allow Alice to configure x509 and Bob to configure OIDC" Is that kind of case just delegated to admission?

What about persistant connections? If I remove an x509 config, and there is a client using a cert from that CA with an open connection, what happens?

Reply:
> This new API is obviously admin-level.

Yeah the assumption is cluster-admin level access.

> Are there any gaps in RBAC to managing this API? Ex: "I only want to allow Alice to configure x509 and Bob to configure OIDC"

While we could certainly build custom permission checks per `spec.type`, I am not sure that would be meaningful since you can assert basically any identity from x509 or OIDC.  You would just manufacture an identity that had enough authz to access the other `spec.type`.

> Is that kind of case just delegated to admission?

For super custom stuff, yes.  But we have plenty of flexibility via the REST storage layer to add restrictions if we deem them to be valuable.

> Is our only answer to preventing privilege escalation through the API "use RBAC and Admission"?

I think that is generally true for such APIs, though if there are ways we can add meaningful protection directly built-in to k/k, then I am all for that instead of asking people to write custom admission webhooks.

> I just want to be conscious that we're introducing a new privilege escalation vector.

I am not against following a similar pattern to RBAC's cluster role aggregation, i.e. requiring full cluster-admin access:

```go
var fullAuthority = []rbac.PolicyRule{
	rbac.NewRule("*").Groups("*").Resources("*").RuleOrDie(),
	rbac.NewRule("*").URLs("*").RuleOrDie(),
}

func (s *Storage) Create(ctx context.Context, obj runtime.Object, createValidatingAdmission rest.ValidateObjectFunc, options *metav1.CreateOptions) (runtime.Object, error) {
	if rbacregistry.EscalationAllowed(ctx) || rbacregistry.RoleEscalationAuthorized(ctx, s.authorizer) {
		return s.StandardStorage.Create(ctx, obj, createValidatingAdmission, options)
	}

	clusterRole := obj.(*rbac.ClusterRole)

	// ...

	// to set the aggregation rule, since it can gather anything, requires * on *.*
	if hasAggregationRule(clusterRole) {
		if err := rbacregistryvalidation.ConfirmNoEscalationInternal(ctx, s.ruleResolver, fullAuthority); err != nil {
			return nil, apierrors.NewForbidden(groupResource, clusterRole.Name, errors.New("must have cluster-admin privileges to use the aggregationRule"))
		}
	}

	return s.StandardStorage.Create(ctx, obj, createValidatingAdmission, options)
}
```

> What about persistant connections? If I remove an x509 config, and there is a client using a cert from that CA with an open connection, what happens?

I am not 100% sure on this one but this problem already exists today since we dynamically reload CA bundles.  I _think_ we close connections.

Reply:
Note also that if you have cluster-admin level access, you can just run an impersonation proxy / use impersonation to create any identity, similar to this API.